pacman/model/graphs/application/abstract/abstract_2d_device_vertex.py

```diff
@@ -0,0 +1,248 @@
+# Copyright (c) 2017-2019 The University of Manchester
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from spinn_utilities.abstract_base import AbstractBase, abstractproperty
+import math
+from pacman.exceptions import PacmanConfigurationException
+from pacman.utilities.utility_calls import get_n_bits
+from pacman.utilities.constants import BITS_IN_KEY
+from pacman.model.routing_info.base_key_and_mask import BaseKeyAndMask
+from pacman.model.graphs.common.slice import Slice
+
+
+class Abstract2DDeviceVertex(object, metaclass=AbstractBase):
+    """ A helper for 2D input devices.  Note that this assumes that
+        the input keys will contain a field for each of the X and Y dimensions
+        with X field in the LSBs and the Y field in the next adjacent bits.  If
+        the fields are in different places, override the methods:
+        _source_x_shift, _source_y_shift, _source_x_mask and _source_y_mask.
+        If the key has bits in addition to the X and Y values, you can also
+        override _key_shift.
+    """
+
+    @abstractproperty
+    def _width(self):
```

Comment:
These properties aren't necessarily meant for the outside world, though I can see that they would cause no issues.  They are really there so that the abstract doesn't need slots of its own.